src/main.cpp

```diff
@@ -2871,6 +2883,22 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         }
     }
 
+    // History read/write is unconditional, since it is generated for entire
+    // chain regardless of the Heartwood upgrade activation height.
+    {
+        auto historyNode = libzcash::NewLeaf(
```

Comment:
The PR does not yet alter the consensus rules on the block header to require that it contain the correct history root (for the previous block's history tree) instead of the correct Sapling anchor (for this block).

Reply:
I added condition=false which should be replaced with heartwood update code, which is out of scope of this pr